### PR TITLE
Add high level communication module of chargebyte

### PR DESCRIPTION
**Pull request by mooraby**
This pull request adds a new module named EvseV2G. This module provides a C/C++ based implementation of the ISO 15118/DIN 70121 EIM standard for charging stations. It implements the “ISO15118_charger” interface of the EVerest framework. This module is contributed by chargebyte GmbH.

Implemented features:

- Support of ISO 15118-2 EIM AC and DC
- Support of DIN 70121 EIM
- Prototypical implementation of ISO 15118-2 PnC

### CODE REVIEW

File: modules/EvseV2G/EvseV2G.cpp [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    int rv = 0;
+    /* create v2g context */
+    v2g_ctx = v2g_ctx_create(&(*p_charger));
+
+    if (v2g_ctx == NULL)
+        return;
+
+    invoke_init(*p_charger);
+
+    dlog(DLOG_LEVEL_INFO, "Starting SDP responder");
```
Comment (SebaLukas) [inline modules/EvseV2G/EvseV2G.cpp:23]: Debug instead of log

File: modules/EvseV2G/iso_server.cpp [hunk 2]
```diff
@@ -0,0 +1,2222 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright (C) 2023 chargebyte GmbH
+// Copyright (C) 2023 Contributors to EVerest
+
+#include <inttypes.h>
+#include <json-c/json.h>
```
Comment (corneliusclaussen) [inline modules/EvseV2G/iso_server.cpp:6]: Adds a dependency to libjson-c, but does not use it. Suggested to remove include

File: modules/EvseV2G/v2g_server.cpp [hunk 3]
```diff
@@ -0,0 +1,614 @@
...
+#include "connection.hpp"
+#include "din_server.hpp"
+#include "iso_server.hpp"
+#include "log.hpp"
+#include "tools.hpp"
+#include "v2g_server.hpp"
+
+#define MAX_RES_TIME 98
+
+static types::iso15118_charger::V2G_Message_ID get_V2G_Message_ID(enum V2gMsgTypeId v2g_msg, bool is_req) {
```
Comment (SebaLukas) [inline modules/EvseV2G/v2g_server.cpp:29]: DIN specific messages are missing. E.g. ContractAuthentication

File: README.md [hunk 4]
```diff
@@ -88,6 +88,17 @@ cd ~/checkout/everest-workspace/Josev
...
+
+Change the directory and install openv2g which is a dependency for ISO15118 module EvseV2G:
+```bash
+cd ~/checkout/everest-workspace/openv2g
+mkdir build
+cd build
+cmake ..
+sudo make install
+```
+
```
Comment (SebaLukas) [inline README.md:101]: No longer relevant when openv2g is used as fork

**Review by SebaLukas — CHANGES_REQUESTED** [review]
I went over the code roughly and didn't find anything big. Please have a look at my comments.
There are still a few codacy errors. But they should be relatively easy to fix.

The libevent-devel and the json-c (if it is really needed) package was missing. Please add these also in the Readme.md.

File: modules/EvseV2G/charger/ISO15118_chargerImpl.cpp [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+        v2g_ctx->debugMode = false;
+    } else {
+        v2g_ctx->debugMode = true;
+    }
+};
+
+void ISO15118_chargerImpl::handle_set_Auth_Okay_EIM(bool& auth_okay_eim) {
+    if (auth_okay_eim == true) {
+        v2g_ctx->evse_v2g_data.evse_processing[PHASE_AUTH] = (uint8_t)iso1EVSEProcessingType_Finished;
+    }
```
Comment (SebaLukas) [inline modules/EvseV2G/charger/ISO15118_chargerImpl.cpp:174]: Where is the else case?

File: modules/EvseV2G/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+target_sources(${MODULE_NAME}
+    PRIVATE
+        v2g_ctx.cpp
+        log.cpp
+        sdp.cpp
+        connection.cpp
+        tools.cpp
+        v2g_server.cpp
+        iso_server.cpp
+        din_server.cpp
```
Comment (SebaLukas) [inline modules/EvseV2G/CMakeLists.txt:45]: Please change to:
```cpp
"v2g_ctx.cpp"
"log.cpp"
"sdp.cpp"
"connection.cpp"
"tools.cpp"
"v2g_server.cpp"
"iso_server.cpp"
"din_server.cpp"
```

File: dependencies.yaml [hunk 7]
```diff
@@ -43,3 +43,12 @@ libocpp:
...
   git: https://github.com/EVerest/ext-switchev-iso15118.git
   git_tag: 2023.1.0
+
+# OpenV2G
+ext-openv2g:
+  git: https://github.com/EVerest/ext-openv2g.git
+  git_tag: everest
+# mbedtls
+libmbedtls:
+  git: https://github.com/chargebyte/mbedtls.git
```
Comment (SebaLukas) [inline dependencies.yaml:53]: Please change to:
`git: https://github.com/EVerest/ext-mbedtls.git`

This is a fork from your mbedtls repo.

**Review by SebaLukas — CHANGES_REQUESTED** [review]
I found three more things that should be changed.
After the changes then also please do a squash :)

**Comment by SebaLukas** [comment]
I will look for the codacy issues in parallel. And try to fix them.

File: modules/EvseV2G/charger/ISO15118_chargerImpl.cpp (hunk 5 above)
Comment (SebaLukas) [reply modules/EvseV2G/charger/ISO15118_chargerImpl.cpp:174]: Alright :thumbsup:

File: modules/EvseV2G/connection.cpp [hunk 8]
```diff
@@ -0,0 +1,946 @@
...
+        break;
+    }
+}
+
+static bool connection_init_tls(struct v2g_context* ctx) {
+
+    int rv;
+
+    char* file_path =
+        static_cast<char*>(malloc(max(strlen(DEFAULT_PKI_PATH), strlen(DEFAULT_KEY_PATH)) + MAX_FILE_NAME_LENGTH));
```
Comment (SebaLukas) [inline modules/EvseV2G/connection.cpp:432]:
```cpp
static_cast<char*>(malloc(max(std::string(DEFAULT_PKI_PATH).size(), std::string(DEFAULT_KEY_PATH).size()) + MAX_FILE_NAME_LENGTH));
```

File: modules/EvseV2G/connection.cpp [hunk 9]
```diff
@@ -0,0 +1,946 @@
...
+    DIR* d = opendir(DEFAULT_PKI_PATH); // open the path
+    struct dirent* dir;                 // for the directory entries
+
+    if (d == NULL) {
+        dlog(DLOG_LEVEL_ERROR, "Unable to open pki file path: %s", DEFAULT_PKI_PATH);
+        goto error_out;
+    }
+
+    while ((dir = readdir(d)) != NULL)
+        if ((dir->d_type == DT_DIR) && (strlen(dir->d_name) && isdigit((int)dir->d_name[0])) &&
```
Comment (SebaLukas) [inline modules/EvseV2G/connection.cpp:448]:
```cpp
if ((dir->d_type == DT_DIR) && (std::string(dir->d_name).size() && isdigit((int)dir->d_name[0])) &&
```

File: modules/EvseV2G/connection.cpp [hunk 10]
```diff
@@ -0,0 +1,946 @@
...
+    /* cleanup and notify lower layers */
+    connection_teardown(conn);
+
+    free(conn);
+
+    return NULL;
+}
+
+static void* connection_server(void* data) {
+    struct v2g_context* ctx = (struct v2g_context*)data;
```
Comment (SebaLukas) [inline modules/EvseV2G/connection.cpp:837]:
```cpp
struct v2g_context* ctx =  static_cast<v2g_context*>(data);
```

File: modules/EvseV2G/iso_server.cpp [hunk 11]
```diff
@@ -0,0 +1,2233 @@
...
+static bool publish_iso_payment_details_req(struct iso1PaymentDetailsReqType const* const v2g_payment_details_req) {
+
+    bool rv = true;
+    unsigned char* base64Buffer = NULL;
+    size_t olen;
+
+    /* Parse contract leaf certificate */
+    if (MAX_CERT_SIZE >= v2g_payment_details_req->ContractSignatureCertChain.Certificate.bytesLen) {
+        mbedtls_base64_encode(NULL, 0, &olen,
+                              (unsigned char*)v2g_payment_details_req->ContractSignatureCertChain.Certificate.bytes,
```
Comment (SebaLukas) [inline modules/EvseV2G/iso_server.cpp:570]: `v2g_payment_details_req->ContractSignatureCertChain.Certificate.bytes` without casting

File: modules/EvseV2G/iso_server.cpp [hunk 12]
```diff
@@ -0,0 +1,2233 @@
...
+    // TODO: add leaf certificate to contract chain: (char*) base64Buffer, olen
+
+    /* Parse contract sub certificate */
+    if (v2g_payment_details_req->ContractSignatureCertChain.SubCertificates_isUsed == (unsigned int)1) {
+        const struct iso1SubCertificatesType* subCertTmp =
+            &v2g_payment_details_req->ContractSignatureCertChain.SubCertificates;
+
+        for (uint8_t idx = 0; idx < subCertTmp->Certificate.arrayLen; idx++) {
+            if (subCertTmp->Certificate.array[idx].bytesLen <= MAX_CERT_SIZE) {
+                mbedtls_base64_encode(NULL, 0, &olen, (unsigned char*)subCertTmp->Certificate.array[idx].bytes,
```
Comment (SebaLukas) [inline modules/EvseV2G/iso_server.cpp:599]: `mbedtls_base64_encode(NULL, 0, &olen, subCertTmp->Certificate.array[idx].bytes,`

File: modules/EvseV2G/iso_server.cpp [hunk 13]
```diff
@@ -0,0 +1,2233 @@
...
+                                      (size_t)subCertTmp->Certificate.array[idx].bytesLen);
+                if (base64Buffer != NULL)
+                    free(base64Buffer);
+                base64Buffer = (unsigned char*)malloc(olen);
```
Comment (SebaLukas) [inline modules/EvseV2G/iso_server.cpp:603]: `base64Buffer = static_cast<unsigned char*>(malloc(olen));`

File: modules/EvseV2G/iso_server.cpp [hunk 14]
```diff
@@ -0,0 +1,2233 @@
...
+            } else {
+                dlog(DLOG_LEVEL_ERROR, "Invalid size of the received contract sub certificate #%d", idx);
+                rv = false;
+                goto exit;
+            }
+
+            if ((NULL == base64Buffer) ||
+                (0 != mbedtls_base64_encode(base64Buffer, olen, &olen,
+                                            (unsigned char*)subCertTmp->Certificate.array[idx].bytes,
```
Comment (SebaLukas) [inline modules/EvseV2G/iso_server.cpp:612]: `subCertTmp->Certificate.array[idx].bytes,` without casting

File: modules/EvseV2G/iso_server.cpp [hunk 15]
```diff
@@ -0,0 +1,2233 @@
...
+        sprintf(&buffer[i * 3], "%02" PRIx8 ":", req->EVCCID.bytes[i]);
+    }
+    if (i)
+        buffer[i * 3 - 1] = '\0';
+    else
+        buffer[0] = '\0';
+
+    conn->ctx->p_charger->publish_EVCCIDD(buffer); // publish EVCC ID
+
+    dlog(DLOG_LEVEL_INFO, "SessionSetupReq.EVCCID: %s", strlen(buffer) ? buffer : "(zero length provided)");
```
Comment (SebaLukas) [inline modules/EvseV2G/iso_server.cpp:800]: `dlog(DLOG_LEVEL_INFO, "SessionSetupReq.EVCCID: %s", std::string(buffer).size() ? buffer : "(zero length provided)");`

File: modules/EvseV2G/iso_server.cpp [hunk 16]
```diff
@@ -0,0 +1,2233 @@
...
+                req->eMAID.characters[emaid_len++] = req->eMAID.characters[idx];
+            }
+        }
+
+        req->eMAID.charactersLen = emaid_len;
+        req->eMAID.characters[emaid_len] = '\0';
+
+        dlog(DLOG_LEVEL_TRACE, "emaid-v2g: %s emaid-cert: %s", req->eMAID.characters, cert_emaid);
+
+        if ((strlen(cert_emaid) != req->eMAID.charactersLen) ||
```
Comment (SebaLukas) [inline modules/EvseV2G/iso_server.cpp:1139]: `if ((std::string(cert_emaid).size() != req->eMAID.charactersLen) ||`

File: modules/EvseV2G/iso_server.cpp [hunk 17]
```diff
@@ -0,0 +1,2233 @@
...
+
+        // Convert the public key in the certificate to a mbed TLS ECDSA public key
+        // This also verifies that it's an ECDSA key and not an RSA key
+        err = mbedtls_ecdsa_from_keypair(&conn->ctx->session.contract.pubkey,
+                                         mbedtls_pk_ec(conn->ctx->session.contract.crt.pk));
+        if (err != 0) {
+            memset(res, 0, sizeof(*res));
+            res->ResponseCode = iso1responseCodeType_FAILED_CertChainError;
+            char strerr[256];
+            mbedtls_strerror(err, strerr, strlen(strerr));
```
Comment (SebaLukas) [inline modules/EvseV2G/iso_server.cpp:1161]: `mbedtls_strerror(err, strerr, std::string(strerr).size());`

File: modules/EvseV2G/iso_server.cpp [hunk 18]
```diff
@@ -0,0 +1,2233 @@
...
+            dlog(DLOG_LEVEL_ERROR, "CertificateInstallationRes timeout occured");
+            conn->ctx->intl_emergency_shutdown = true; // [V2G2-918] Initiating emergency shutdown, response code faild
+                                                       // will be set in iso_validate_response_code() function
+        }
+        pthread_mutex_unlock(&conn->ctx->mqtt_lock);
+    }
+    if (conn->ctx->evse_v2g_data.cert_install_res_b64_buffer != NULL) {
+        if ((rv = mbedtls_base64_decode(conn->buffer + V2GTP_HEADER_LENGTH, DEFAULT_BUFFER_SIZE, &conn->buffer_pos,
+                                       reinterpret_cast<unsigned char*>(conn->ctx->evse_v2g_data.cert_install_res_b64_buffer),
+                                       strlen(conn->ctx->evse_v2g_data.cert_install_res_b64_buffer))) != 0) {
```
Comment (SebaLukas) [inline modules/EvseV2G/iso_server.cpp:1770]: `std::string(conn->ctx->evse_v2g_data.cert_install_res_b64_buffer).size())) != 0) {`

File: modules/EvseV2G/log.cpp [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+    snprintf(log_datetimestamp + offset, sizeof(log_datetimestamp) - offset, ".%03ld] ", debug_tval.tv_usec / 1000);
+
+    va_list args;
+    va_start(args, format);
+
+    // print the user given part
+    // strip possible newline character from user-given string
+    // FIXME: could be skipped
+    if (format) {
+        size_t formatlen = strlen(format);
```
Comment (SebaLukas) [inline modules/EvseV2G/log.cpp:53]: `size_t formatlen = std::string(format).size();`

File: modules/EvseV2G/tools.cpp [hunk 20]
```diff
@@ -0,0 +1,330 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright (C) 2022-2023 chargebyte GmbH
+// Copyright (C) 2022-2023 Contributors to EVerest
+#include <arpa/inet.h>
+#include <dirent.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <ifaddrs.h>
+#include <math.h>
+#include <stdio.h>
```
Comment (SebaLukas) [inline modules/EvseV2G/tools.cpp:10]: Adding here `#include <string>`

File: modules/EvseV2G/tools.cpp [hunk 21]
```diff
@@ -0,0 +1,330 @@
...
+        return false;
+    }
+    DIR* d = opendir(path); // open the path
+
+    if (d == NULL) {
+        dlog(DLOG_LEVEL_ERROR, "Unable to open file path %s", path);
+        return false;
+    }
+    struct dirent* dir; // for the directory entries
+    uint8_t file_name_identifier_len = strlen(file_name_identifier);
```
Comment (SebaLukas) [inline modules/EvseV2G/tools.cpp:252]: `uint8_t file_name_identifier_len = std::string(file_name_identifier).size();`

File: modules/EvseV2G/tools.cpp [hunk 22]
```diff
@@ -0,0 +1,330 @@
...
+    while ((dir = readdir(d)) != NULL) {
+        if (dir->d_type != DT_DIR) {
+            /* if the type is not directory*/
+            if ((strlen(dir->d_name) > (file_name_identifier_len)) && /* Plus one for the numbering */
+                (strncmp(file_name_identifier, dir->d_name, file_name_identifier_len) == 0) &&
+                (file_name_len > strlen(dir->d_name))) {
+                strncpy(file_name, dir->d_name, strlen(dir->d_name) + 1);
+                break;
+            }
```
Comment (SebaLukas) [inline modules/EvseV2G/tools.cpp:261]:
```cpp
if ((std::string(dir->d_name).size() > (file_name_identifier_len)) && /* Plus one for the numbering */
    (strncmp(file_name_identifier, dir->d_name, file_name_identifier_len) == 0) &&
    (file_name_len > std::string(dir->d_name).size())) {
    strncpy(file_name, dir->d_name, std::string(dir->d_name).size() + 1);
    break;
}
```

File: modules/EvseV2G/tools.cpp [hunk 23]
```diff
@@ -0,0 +1,330 @@
...
+    DIR* d = opendir(path); // open the path
+
+    if (d == NULL) {
+        dlog(DLOG_LEVEL_ERROR, "Unable to open file path %s", path);
+        return 0;
+    }
+    struct dirent* dir; // for the directory entries
+    uint8_t num_of_files = 0;
+    uint8_t prefix_len = strlen(prefix);
+    uint8_t suffix_len = strlen(suffix);
```
Comment (SebaLukas) [inline modules/EvseV2G/tools.cpp:287]:
```cpp
uint8_t prefix_len = std::string(prefix).size();
uint8_t suffix_len = std::string(suffix).size();
```

File: modules/EvseV2G/tools.cpp [hunk 24]
```diff
@@ -0,0 +1,330 @@
...
+    uint8_t min_idx = max_idx; // helper value to re-sort array
+
+    while (((dir = readdir(d)) != NULL) && (num_of_files != (max_idx - offset))) {
+        if (dir->d_type != DT_DIR) {
+            /* if the type is not directory*/
+            if ((strlen(dir->d_name) > (prefix_len + suffix_len + 1)) && /* Plus one for the numbering */
```
Comment (SebaLukas) [inline modules/EvseV2G/tools.cpp:293]: `if ((std::string(dir->d_name).size() > (prefix_len + suffix_len + 1)) && /* Plus one for the numbering */`

File: modules/EvseV2G/tools.cpp [hunk 25]
```diff
@@ -0,0 +1,330 @@
...
+                (0 == strncmp(prefix, dir->d_name, prefix_len))) {
+                for (uint8_t idx = offset; idx < max_idx; idx++) {
+                    /* Iterated over the number prefix */
+                    if ((dir->d_name[prefix_len] == ('0' + idx)) &&
+                        (strncmp(suffix, &dir->d_name[prefix_len + 1], suffix_len) == 0)) {
+                        if (MAX_FILE_NAME_LENGTH >= strlen(dir->d_name)) {
```
Comment (SebaLukas) [inline modules/EvseV2G/tools.cpp:299]: `if (MAX_FILE_NAME_LENGTH >= std::string(dir->d_name).size()) {`

File: modules/EvseV2G/v2g_ctx.cpp [hunk 26]
```diff
@@ -0,0 +1,535 @@
...
+    // ctx->evse_v2g_data.chargeService.ServiceScope.charactersLen
+    ctx->evse_v2g_data.charge_service.ServiceScope_isUsed = (unsigned int)0;
+
+    ctx->meter_info.meter_info_is_used = false;
+
+    if (initialize_once == false) {
+        ctx->evse_v2g_data.charge_service.FreeService = 0;
+        strcpy((char*)ctx->evse_v2g_data.evse_id.bytes, "DE*CBY*ETE1*234");
+        ctx->evse_v2g_data.evse_id.bytesLen =
+            static_cast<uint16_t>(strlen((const char*)ctx->evse_v2g_data.evse_id.bytes));
```
Comment (SebaLukas) [inline modules/EvseV2G/v2g_ctx.cpp:170]:
```cpp
std::string evse_id = std::string("DE*CBY*ETE1*234");
strcpy(reinterpret_cast<char*>(ctx->evse_v2g_data.evse_id.bytes), evse_id.data());
ctx->evse_v2g_data.evse_id.bytesLen = evse_id.size();
```

File: modules/EvseV2G/v2g_ctx.cpp [hunk 27]
```diff
@@ -0,0 +1,535 @@
...
+
+        parameterSet->ParameterSetID = parameterSetId;
+
+        /* Service to install a Contract Certificate (Ref. Table 106 —
+         * ServiceParameterList for certificate service) */
+        if (parameterSet->ParameterSetID == 1) {
+            /* Configure parameter name */
+            strcpy(parameterSet->Parameter.array[0].Name.characters, "Service");
+            parameterSet->Parameter.array[0].Name.charactersLen =
+                strlen(parameterSet->Parameter.array[0].Name.characters);
```
Comment (SebaLukas) [inline modules/EvseV2G/v2g_ctx.cpp:506]: `std::string(parameterSet->Parameter.array[0].Name.characters).size();`

File: modules/EvseV2G/v2g_ctx.cpp [hunk 28]
```diff
@@ -0,0 +1,535 @@
...
+
+            /* Configure parameter value */
+            strcpy(parameterSet->Parameter.array[0].stringValue.characters, "Installation");
+            parameterSet->Parameter.array[0].stringValue.charactersLen =
+                strlen(parameterSet->Parameter.array[0].stringValue.characters);
```
Comment (SebaLukas) [inline modules/EvseV2G/v2g_ctx.cpp:511]: `std::string(parameterSet->Parameter.array[0].stringValue.characters).size();`

File: modules/EvseV2G/v2g_ctx.cpp [hunk 29]
```diff
@@ -0,0 +1,535 @@
...
+            parameterSet->Parameter.array[0].stringValue_isUsed = 1;
+            parameterSet->Parameter.arrayLen = 1;
+        }
+        /* Service to update a Contract Certificate */
+        else if (parameterSet->ParameterSetID == 2) {
+            /* Configure parameter name */
+            strcpy(parameterSet->Parameter.array[0].Name.characters, "Service");
+            parameterSet->Parameter.array[0].Name.charactersLen =
+                strlen(parameterSet->Parameter.array[0].Name.characters);
```
Comment (SebaLukas) [inline modules/EvseV2G/v2g_ctx.cpp:520]: `std::string(parameterSet->Parameter.array[0].Name.characters).size();`

File: modules/EvseV2G/v2g_ctx.cpp [hunk 30]
```diff
@@ -0,0 +1,535 @@
...
+
+            /* Configure parameter value */
+            strcpy(parameterSet->Parameter.array[0].stringValue.characters, "Update");
+            parameterSet->Parameter.array[0].stringValue.charactersLen =
+                strlen(parameterSet->Parameter.array[0].stringValue.characters);
```
Comment (SebaLukas) [inline modules/EvseV2G/v2g_ctx.cpp:525]: `std::string(parameterSet->Parameter.array[0].stringValue.characters).size();`

File: modules/EvseV2G/v2g_server.cpp [hunk 31]
```diff
@@ -0,0 +1,629 @@
...
+ */
+static void publish_var_V2G_Message(v2g_connection* conn, bool is_req) {
+    types::iso15118_charger::V2G_Messages v2gMessage;
+
+    u_int8_t* tempbuff = conn->buffer;
+    std::string msg_as_hex_string;
+    for (int i = 0; ((tempbuff != NULL) && (i < conn->payload_len + V2GTP_HEADER_LENGTH)); i++) {
+        char hex[4];
+        snprintf(hex, 4, "%x", *tempbuff); // to hex
+        if (strlen(hex) == 1)
```
Comment (SebaLukas) [inline modules/EvseV2G/v2g_server.cpp:111]: `if (std::string(hex).size() == 1)`

File: modules/EvseV2G/v2g_server.cpp [hunk 32]
```diff
@@ -0,0 +1,629 @@
...
+            msg_as_hex_string += '0';
+        msg_as_hex_string += hex;
+        tempbuff++;
+    }
+
+    unsigned char* base64_buffer = NULL;
+    size_t base64_buffer_len = 0;
+    mbedtls_base64_encode(NULL, 0, &base64_buffer_len, (unsigned char*)conn->buffer,
```
Comment (SebaLukas) [inline modules/EvseV2G/v2g_server.cpp:119]: Casting is not needed

File: modules/EvseV2G/v2g_server.cpp [hunk 33]
```diff
@@ -0,0 +1,629 @@
...
+                          (size_t)conn->payload_len + V2GTP_HEADER_LENGTH);
+    base64_buffer = static_cast<unsigned char*>(malloc(base64_buffer_len));
+    if ((base64_buffer == NULL) || (mbedtls_base64_encode(base64_buffer, base64_buffer_len, &base64_buffer_len,
+                                                          static_cast<unsigned char*>(conn->buffer),
+                                                          (size_t)conn->payload_len + V2GTP_HEADER_LENGTH) != 0)) {
+        dlog(DLOG_LEVEL_WARNING, "Unable to base64 encode EXI buffer");
+    }
+
+    v2gMessage.V2G_Message_EXI_Base64 = std::string((const char*)base64_buffer, base64_buffer_len);
```
Comment (SebaLukas) [inline modules/EvseV2G/v2g_server.cpp:128]: `v2gMessage.V2G_Message_EXI_Base64 = std::string(reinterpret_cast<char const*>(base64_buffer), base64_buffer_len);`

File: modules/EvseV2G/v2g_server.cpp [hunk 34]
```diff
@@ -0,0 +1,629 @@
...
+
+int v2g_handle_connection(struct v2g_connection* conn) {
+    int rv = -1;
+    enum v2g_event rvAppHandshake = V2G_EVENT_NO_EVENT;
+    bool stop_receiving_loop = false;
+    int64_t start_time = 0; // in ms
+
+    enum v2g_protocol selected_protocol = V2G_UNKNOWN_PROTOCOL;
+    v2g_ctx_init_charging_state(conn->ctx, false);
+    conn->buffer = (uint8_t*)malloc(DEFAULT_BUFFER_SIZE);
```
Comment (SebaLukas) [inline modules/EvseV2G/v2g_server.cpp:346]: `conn->buffer = static_cast<uint8_t*>(malloc(DEFAULT_BUFFER_SIZE));`

File: modules/EvseV2G/v2g_server.cpp [hunk 35]
```diff
@@ -0,0 +1,629 @@
...
+                break;
+            default:
+                goto error_out; //     if protocol is unknown
+            }
+            /* Wait max. res-time before sending the next response */
+            int64_t time_to_conf_res = getmonotonictime() - start_time;
+
+            if (time_to_conf_res < MAX_RES_TIME) {
+                // dlog(DLOG_LEVEL_ERROR,"time_to_conf_res %llu", time_to_conf_res);
+                usleep((MAX_RES_TIME - time_to_conf_res) * 1000);
```
Comment (SebaLukas) [inline modules/EvseV2G/v2g_server.cpp:547]:
```cpp
std::this_thread::sleep_for(
    std::chrono::microseconds((MAX_RES_TIME - time_to_conf_res) * 1000));
```

**Review by SebaLukas — CHANGES_REQUESTED** [review]
Codacy issues fixes

**Comment by SebaLukas** [comment]
Please use squash and merge :)

**Pull request merged**